// js-unpacker/js_unpacker.js
/**
 * Clase JsUnpacker
 * 
 * Esta clase se utiliza para detectar y deofuscar código JavaScript ofuscado utilizando el método P.A.C.K.E.R.
 */
class JsUnpacker {
  /**
   * Constructor de JsUnpacker.
   *
   * @param {string} packedJS - El código JavaScript ofuscado.
   */
  constructor(packedJS) {
      this.packedJS = packedJS;
  }

  /**
   * Detecta si el código JavaScript está ofuscado utilizando el método P.A.C.K.E.R.
   *
   * @returns {boolean} - Devuelve true si el código está ofuscado con P.A.C.K.E.R., de lo contrario false.
   */
  detect() {
      const js = this.packedJS.replace(/\s+/g, "");
      const pattern = /eval\(function\(p,a,c,k,e,(?:r|d)/;
      return pattern.test(js);
  }

  /**
   * Deofusca el código JavaScript ofuscado.
   *
   * @returns {string|null} - El código JavaScript deofuscado, o null si no se pudo deofuscar.
   */
  unpack() {
      let js = this.packedJS;
      try {
          const pattern = /\}\s*?\('(.*)',\s*(.*?),\s*(\d+),\s*'(.*?)'\.split\('\|'\)/;
          const match = pattern.exec(js);
          if (match && match.length === 5) {
              let payload = match[1].replace(/\\'/g, "'");
              const radixStr = match[2];
              const countStr = match[3];
              const symtab = match[4].split('|');

              let radix = 36;
              let count = 0;
              try {
                  radix = parseInt(radixStr, 10);
              } catch (e) {
                  // El valor por defecto del radix sigue siendo 36
              }
              try {
                  count = parseInt(countStr, 10);
              } catch (e) {
                  // El valor por defecto del count sigue siendo 0
              }

              if (symtab.length !== count) {
                  throw new Error("Unknown p.a.c.k.e.r. encoding");
              }

              const unbase = new Unbase(radix);
              const payloadPattern = /\b\w+\b/g;
              let decoded = payload;
              let replaceOffset = 0;

              let matchPayload;
              while ((matchPayload = payloadPattern.exec(payload)) !== null) {
                  const word = matchPayload[0];
                  const x = unbase.unbase(word);
                  let value = null;
                  if (x < symtab.length) {
                      value = symtab[x];
                  }

                  if (value !== null && value.length > 0) {
                      decoded = decoded.substring(0, matchPayload.index + replaceOffset) + value + decoded.substring(matchPayload.index + replaceOffset + word.length);
                      replaceOffset += (value.length - word.length);
                  }
              }
              return decoded;
          }
      } catch (e) {
          console.error(e);
      }
      return null;
  }
}

/**
* Clase Unbase
* 
* Esta clase se utiliza para convertir cadenas en números en una base específica.
*/
class Unbase {
  /**
   * Constructor de Unbase.
   *
   * @param {number} radix - La base numérica utilizada para la conversión.
   */
  constructor(radix) {
      this.ALPHABET_62 = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
      this.ALPHABET_95 = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
      this.alphabet = null;
      this.dictionnary = new Map();
      this.radix = radix;

      if (radix > 36) {
          if (radix < 62) {
              this.alphabet = this.ALPHABET_62.substring(0, radix);
          } else if (radix > 62 && radix < 95) {
              this.alphabet = this.ALPHABET_95.substring(0, radix);
          } else if (radix === 62) {
              this.alphabet = this.ALPHABET_62;
          } else if (radix === 95) {
              this.alphabet = this.ALPHABET_95;
          }

          for (let i = 0; i < this.alphabet.length; i++) {
              this.dictionnary.set(this.alphabet.charAt(i), i);
          }
      }
  }

  /**
   * Convierte una cadena en un número en la base específica.
   *
   * @param {string} str - La cadena a convertir.
   * @returns {number} - El número convertido.
   */
  unbase(str) {
      let ret = 0;

      if (this.alphabet === null) {
          ret = parseInt(str, this.radix);
      } else {
          const tmp = str.split('').reverse().join('');
          for (let i = 0; i < tmp.length; i++) {
              ret += Math.pow(this.radix, i) * this.dictionnary.get(tmp.charAt(i));
          }
      }
      return ret;
  }
}

module.exports = JsUnpacker;
